bluefog/common/operations.cc

```diff
@@ -186,151 +504,401 @@ bool RunLoopOnce(BluefogGlobalState& state) {
+
+    while (!responses.empty()) {
+      auto response = responses.front();
+      assert(response.tensor_names().size() == 1);
+      responses.pop_front();
+      // TODO: Tensor Fusion Logics.
+      response_list.add_response(response);
+    }
+
+    // Notify all nodes which tensors we'd like to reduce at this step.
+    std::string encoded_response;
+    ResponseList::SerializeToString(response_list, encoded_response);
+    int encoded_response_length = (int)encoded_response.length() + 1;
+    MPI_Bcast(&encoded_response_length, 1, MPI_INT, COORDINATE_RANK,
+              mpi_context.mpi_comm);
+    MPI_Bcast((void*)encoded_response.c_str(), encoded_response_length,
+              MPI_BYTE, COORDINATE_RANK, mpi_context.mpi_comm);
+    // Perform the collective operation. All nodes should end up performing
+    // the same operation.
+    for (auto& response : response_list.responses()) {
+      state.tensor_queue.GetTensorEntriesFromResponse(response, entries);
+      // TODO: tensor fusion logics?
+    }
+
+    // Check for stalled tensors.
+    if (std::chrono::steady_clock::now() - state.last_stall_check >
+        STALL_WARNING_TIME) {
+      CheckForStalledTensors(state);
+      state.last_stall_check = std::chrono::steady_clock::now();
+    }
+  } else {
+    std::string encoded_message;
+    RequestList message_list;
+    message_list.set_shutdown(state.shut_down);
+    message_list.set_change_topo(should_change_topo);
+    while (!message_queue_buffer.empty()) {
+      message_list.add_request(message_queue_buffer.front());
+      message_queue_buffer.pop_front();
+    }
+    RequestList::SerializeToString(message_list, encoded_message);
+    int encoded_message_length = (int)encoded_message.length() + 1;
+    MPI_Gather(&encoded_message_length, 1, MPI_INT, nullptr, 1, MPI_INT,
+               COORDINATE_RANK, mpi_context.mpi_comm);
+    MPI_Gatherv((void*)encoded_message.c_str(), encoded_message_length,
+                MPI_BYTE, nullptr, nullptr, nullptr, MPI_BYTE, COORDINATE_RANK,
+                mpi_context.mpi_comm);
+
+    int msg_length;
+    MPI_Bcast(&msg_length, 1, MPI_INT, COORDINATE_RANK, mpi_context.mpi_comm);
+    auto buffer = new uint8_t[msg_length];
+    MPI_Bcast(buffer, msg_length, MPI_BYTE, COORDINATE_RANK, mpi_context.mpi_comm);
+    ResponseList response_list;
+    ResponseList::ParseFromBytes(response_list, buffer);
+    delete[] buffer;
+
+    // Perform the collective operation. All nodes should end up performing
+    // the same operation.
+    for (auto& response : response_list.responses()) {
+      state.tensor_queue.GetTensorEntriesFromResponse(response, entries);
+      // TODO: tensor fusion logics?
+    }
+
+    if (response_list.shutdown()) {
+      should_shut_down = true;
+    }
+    if (response_list.change_topo()) {
+      should_change_topo = true;
+    }
+  }
+  // Seperate the setting topology and negotiate communnication.
+  if (should_change_topo) {
```

Comment:
Didn't quite understand change_topo logic here and the one at SetSkipNegotiateStageState

Reply:
It is mainly to avoid deadlock. Notice change topo is blocking collective ops and negotiate is also collective blocking ops at two threads. We have to either all participate in the change topo or all doing the negotiation.